reframe/frontend/cli.py

```diff
@@ -183,6 +188,69 @@ def calc_verbosity(site_config, quiesce):
     return curr_verbosity - quiesce
 
 
+def distribute_tests(testcases, skip_system_check, skip_prgenv_check,
+                     node_map):
+    temporary_registry = None
+    new_checks = []
+    for t in testcases:
+        if not t.check.is_fixture():
+            cls = type(t.check)
+            basename = cls.__name__
+            original_var_info = cls.get_variant_info(
+                t.check.variant_num, recurse=True
+            )
+
+            def _rfm_distributed_set_run_nodes(obj):
+                if not obj.local:
+                    obj.job.pin_nodes = obj._rfm_nodelist
+
+            def _rfm_distributed_set_build_nodes(obj):
+                if not obj.local and not obj.build_locally:
+                    obj.build_job.pin_nodes = obj._rfm_nodelist
+
+            # We re-set the valid system and environment in a hook to
+            # make sure that it will not be overwriten by a parent
+            # post-init hook
+            def _rfm_distributed_set_valid_sys_env(obj):
+                obj.valid_systems = [t._partition.fullname]
+                obj.valid_prog_environs = [t._environ.name]
+
+            class BaseTest(t.check.__class__):
+                _rfm_nodelist = builtins.parameter(node_map[t._partition.fullname])
+                valid_systems = [t._partition.fullname]
+                valid_prog_environs = [t._environ.name]
+
+            nc = make_test(
+                f'__D_{t._partition.name}_{t._environ.name}_{basename}',
+                (BaseTest, ),
+                {},
+                methods=[
+                    builtins.run_before('run')(_rfm_distributed_set_run_nodes),
+                    builtins.run_before('compile')(_rfm_distributed_set_build_nodes),
+                    # TODO this hook is not working properly
+                    # builtins.run_after('init')(_rfm_distributed_set_valid_sys_env),
+                ]
+            )
+            # We have to set the prefix manually
+            nc._rfm_dynamic_test_prefix = t.check.prefix
```

Comment:
This should work I think:

```suggestion
            nc._rfm_custom_prefix = t.check.prefix
```